File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.java
class Solution {
    public boolean canPartition(int[] nums) {
        int total_sum =0;
        for(int num:nums)
            total_sum+=num;
        if(total_sum%2!=0){
            return false;
        }
        int subset_sum=total_sum/2;
        boolean dp[] = new boolean[subset_sum+1];
        dp[0] =true;
        for(int num:nums){
            for(int i=subset_sum ;i>=num;i--){
                dp[i]=dp[i]|dp[i-num];
            }
        }
        return dp[subset_sum];
    }
}

File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        
        numSum = sum(nums)
        if numSum%2 !=0:
            return False
        target_sum = numSum//2
        dp = [[-1] * (target_sum + 1) for _ in range(len(nums))]
        
        def f(indx:int ,target:int)->int:
            if nums[indx] == target:
                return True
            if indx == 0:
                return nums[0] == target
            if dp[indx][target] != -1:
                return dp[indx][target]
                
            not_take = f(indx-1,target)
            take = False
            
            if target >= nums[indx]:
                take = f(indx-1,target-nums[indx])
                
            dp[indx][target] = take or not_take
            
            return dp[indx][target]
        
        return f(len(nums)-1,target_sum)